Add a tutorial on mlir-opt (#96105)
llvm/llvm-project: mlir/docs/Tutorials/MlirOpt.md

# Using `mlir-opt`

`mlir-opt` is a command-line entry point for running passes and lowerings on MLIR code.
This tutorial will explain how to use `mlir-opt`, show some examples of its usage,
and mention some useful tips for working with it.

Prerequisites:

- [Building MLIR from source](/getting_started/)
- [MLIR Language Reference](/docs/LangRef/)

[TOC]

## `mlir-opt` basics

The `mlir-opt` tool loads a textual IR or bytecode into an in-memory structure,
and optionally executes a sequence of passes
before serializing back the IR (textual form by default).
It is intended as a testing and debugging utility.

After building the MLIR project,
the `mlir-opt` binary (located in `build/bin`)
is the entry point for running passes and lowerings,
as well as emitting debug and diagnostic data.

Running `mlir-opt` with no flags will consume textual or bytecode IR

> from standard in, parse and run verifiers on it,

Reviewer: ```suggestion
from the standard input, parse and run verifiers on it,
```

> and write the textual format back to standard out.

Reviewer: ```suggestion
and write the textual format back to the standard output.
```

This is a good way to test if an input MLIR is well-formed.

`mlir-opt --help` shows a complete list of flags
(there are nearly 1000).
Each pass has its own flag,
though it is recommended to use `--pass-pipeline`
to run passes rather than bare flags.

## Running a pass

Next we run [`convert-to-llvm`](/docs/Passes/#-convert-to-llvm),
which converts all supported dialects to the `llvm` dialect,
on the following IR:

```mlir
// mlir/test/Examples/mlir-opt/ctlz.mlir
func.func @main(%arg0: i32) -> i32 {
  %0 = math.ctlz %arg0 : i32
  func.return %0 : i32
}
```

After building MLIR, and from the `llvm-project` base directory, run

```bash
build/bin/mlir-opt --pass-pipeline="builtin.module(convert-math-to-llvm)" mlir/test/Examples/mlir-opt/ctlz.mlir
```

which produces

```mlir

> module {

Reviewer: At this point the reader may wonder why the pass added this extra module, but it's coming from the `--no-implicit-module` option that is off by default, so it's actually a parser issue.

Maybe starting with a passthrough invocation could allow to clarify this.

Author: Added explicit modules

mlir/test/Examples/mlir-opt/loop_fusion.mlir
```

```mlir
module {
  func.func @producer_consumer_fusion(%arg0: memref<10xf32>, %arg1: memref<10xf32>) {
    %alloc = memref.alloc() : memref<10xf32>
    %alloc_0 = memref.alloc() : memref<10xf32>
    %cst = arith.constant 0.000000e+00 : f32
    affine.for %arg2 = 0 to 10 {
      affine.store %cst, %alloc[%arg2] : memref<10xf32>
      affine.store %cst, %alloc_0[%arg2] : memref<10xf32>
    }
    affine.for %arg2 = 0 to 10 {
      %0 = affine.load %alloc[%arg2] : memref<10xf32>
      %1 = arith.addf %0, %0 : f32
      affine.store %1, %arg0[%arg2] : memref<10xf32>
    }
    affine.for %arg2 = 0 to 10 {
      %0 = affine.load %alloc_0[%arg2] : memref<10xf32>
      %1 = arith.mulf %0, %0 : f32
      affine.store %1, %arg1[%arg2] : memref<10xf32>
    }
    return
  }
}
```

Options passed to a pass
are specified via the syntax `{option1=value1 option2=value2 ...}`,
i.e., use space-separated `key=value` pairs for each option.

## Building a pass pipeline on the command line

The `--pass-pipeline` flag supports combining multiple passes into a pipeline.
So far we have used the trivial pipeline with a single pass

> that is "anchored" on the `builtin.module` op.

Reviewer: ```suggestion
that is "anchored" on the top-level `builtin.module` op.
```

[Pass anchoring](/docs/PassManagement/#oppassmanager)
is a way for passes to specify
that they only run on particular ops.
While many passes are anchored on `builtin.module`,
if you try to run a pass that is anchored on some other op
inside `--pass-pipeline="builtin.module(pass-name)"`,
it will not run.

Multiple passes can be chained together
by providing the pass names in a comma-separated list
in the `--pass-pipeline` string,
e.g.,
`--pass-pipeline="builtin.module(pass1,pass2)"`.
The passes will be run sequentially.

> To use passes that have nontrivial anchoring,

Reviewer: It's not only "passes that have non-trivial anchoring": it is in general selecting the pass anchoring. That is even if a pass does not have any restriction, running for example `builtin.module(func.func(cse, canonicalize))` would have better threading (and locality) behavior than `builtin.module(cse, canonicalize)`

Author: Added a paragraph to this effect.

    gpu.func @func2(%arg0: i32) -> i32 {
      %0 = arith.addi %arg0, %arg0 : i32
      %1 = arith.addi %arg0, %arg0 : i32
      %2 = arith.addi %0, %1 : i32
      gpu.return %2 : i32
    }
  }
}
```

For a spec of the pass-pipeline textual description language,
see [the docs](/docs/PassManagement/#textual-pass-pipeline-specification).
For more general information on pass management, see [Pass Infrastructure](/docs/PassManagement/#).

## Useful CLI flags

- `--debug` prints all debug information produced by `LLVM_DEBUG` calls.
- `--debug-only="my-tag"` prints only the debug information produced by `LLVM_DEBUG`
  in files that have the macro `#define DEBUG_TYPE "my-tag"`.
  This often allows you to print only debug information associated with a specific pass.
    - `"greedy-rewriter"` only prints debug information
      for patterns applied with the greedy rewriter engine.
    - `"dialect-conversion"` only prints debug information
      for the dialect conversion framework.
 - `--emit-bytecode` emits MLIR in the bytecode format.
 - `--mlir-pass-statistics` print statistics about the passes run.
    These are generated via [pass statistics](/docs/PassManagement/#pass-statistics).

>  - `--mlir-print-ir-after-all` prints the IR after each pass.

Reviewer: ```suggestion
 - `--mlir-print-ir-after-all` prints the IR after each pass,
     and ` --mlir-print-ir-tree-dir` dumps it on disk.
```

The` --mlir-print-ir-tree-dir` is much more practical!

(There is also the `  --mlir-print-ir-before=<pass-arg> ` variant)

Author: > When printing the IR before/after a pass, print file tree rooted at this directory

I can't seem to get this flag to to anything, and the help string doesn't help me understand what it is supposed to do....

Reviewer: Seems like we should improve the help!
It is to be used in conjunction with the other IR printing  options.

For example:
```
$ ./bin/mlir-opt ../mlir/test/Transforms/inlining.mlir  --pass-pipeline="builtin.module(func.func(canonicalize, cse))"  --mlir-print-ir-tree-dir=/tmp/foo --mlir-print-ir-after-all --mlir-print-ir-before-all -o /dev/null

$ find /tmp/foo | sort
/tmp/foo
/tmp/foo/builtin_module_no-symbol-name
/tmp/foo/builtin_module_no-symbol-name/0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/1_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_with_arg
/tmp/foo/builtin_module_no-symbol-name/func_func_func_with_arg/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_with_arg/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_with_arg/0_2_cse.mlir
....
```

Author: Done.

# Using `mlir-opt`

`mlir-opt` is the command-line entry point for running passes and lowerings on MLIR code.
This tutorial will explain how to use `mlir-opt` to run passes, and explain
some details about MLIR's built-in dialects along the way.

Prerequisites:

- [Building MLIR from source](/getting_started/)

[TOC]

## Overview

We start with a brief summary of context that helps to frame
the uses of `mlir-opt` detailed in this article.
For a deeper dive on motivation and design,
see [the MLIR paper](https://arxiv.org/abs/2002.11054).

> Two of the central concepts in MLIR are *dialects* and *lowerings*.

Reviewer: ```suggestion
Two of the central concepts in MLIR are *[dialects](/docs/LangRef/#dialects)* and *lowerings*.
```

Not sure that "lowering" is really a concept that needs to be introduced in a mlir-opt documentation.

but each with a set of types, operations, metadata,
and semantics that defines what the operations do.
Different dialects may coexist in the same program.
Then, one writes a set of *lowering passes*
that incrementally converts different parts of the program
from higher level dialects to lower and lower dialects
until you get to machine code
(or, in many cases, LLVM, which finishes the job).
Along the way,
*optimizing passes* are run to make the code more efficient.
The main point here is that the high level dialects exist
*so that* they make it easy to write these important optimizing passes.

A central motivation for building MLIR
was to build the `affine` dialect,
which is designed to enable [polyhedral optimizations](https://polyhedral.info/)
for loop transformations.
Compiler engineers had previously implemented polyhedral optimizations
in LLVM and GCC (without an `affine` dialect),
and it was difficult because they had to reconstruct well-structured loop nests
from a much more complicated set of low-level operations.
Having a higher level `affine` dialect preserves the loop nest structure
at an abstraction layer that makes it easier to write optimizations,

> and then discards it during lowering passes.

Reviewer: I'm not convinced that the two previous paragraphs are desirable here. We can keep it simple I believe:

```
Compilers are operating on a representation of the program, commonly
referred to as "IR" (Intermediate Representation). To keep the compiler
testable and understandable, in general we organize transformations as
separate "passes". A pass takes some IR as input, and mutate it to produce
a new IR.

In MLIR, the IR exists under 3 forms that are equivalent: in-memory, textual,
and bytecode. The in-memory format is the data-structure that the passes
are operating on, the textual format is used for debugging and testing, and
the bytecode for general serialization. MLIR allows to convert from any form
to another.
```

Even this seems like it belongs to a general tutorial more than a MLIR-opt one.

Reviewer: I guess I should have read the description of the PR better, I anchored too much to the title? Overall this still seems like it mixes up things to me and won't mesh with the rest of the documentation by trying to be "self-contained".

> The `mlir-opt` tool can run both
> optimization passes and lowerings,
> though the final code generation
> is performed by a different tool called `mlir-translate`.
> In particular, `mlir-opt` consumes MLIR as input and produce MLIR as output,
> while `mlir-translate` consumes MLIR as input
> and produces non-MLIR program representations as output.
> 

Reviewer: ```suggestion
The `mlir-opt` tool is loading the textual or bytecode form into the in-memory
structure and optionally execute a sequence of passes before serializing back
the IR (textual form by default). It is intended as a testing and debugging utility.
```

Author: > It is intended as a testing and debugging utility

This makes it sound like production use cases do not invoke `mlir-opt` directly, but surely that can't be true... You mean to say that they build their own `mlir-opt`-like tool and invoke that instead?

Reviewer: > This makes it sound like production use cases do not invoke mlir-opt directly, but surely that can't be true... 

This is exactly what I am saying actually :)

## Two example programs

> Here are two MLIR programs.

Reviewer: ```suggestion
Here are the textual representation of two MLIR programs.
```

using the [`math` dialect's](/docs/Dialects/MathOps/) `ctlz` operation.

```mlir
// mlir/test/Examples/mlir-opt/ctlz.mlir
func.func @main(%arg0: i32) -> i32 {
  %0 = math.ctlz %arg0 : i32
  func.return %0 : i32
}
```

This shows the basic structure of an MLIR operation
([see here](https://mlir.llvm.org/docs/LangRef/#operations) for a more complete spec).
Variable names are prefixed with `%`,
functions by `@`,
and each variable/value in a program has a type,
often expressed after a colon.
In this case all the types are `i32`,
except for the function type which is `(i32) -> i32`
(not specified explicitly above, but you'll see it in the `func.call` later).

Each statement is anchored around an expression like `math.ctlz`
which specifies the dialect [`math`](https://mlir.llvm.org/docs/Dialects/MathOps/) via a namespace,
and the operation [`ctlz`](https://mlir.llvm.org/docs/Dialects/MathOps/#mathctlz-mathcountleadingzerosop) after the `.`.
The rest of the syntax of the operation
is determined by a parser defined by the dialect,
and so many operations will have different syntaxes.
In the case of `math.ctlz`,
the sole argument is an integer whose leading zeros are to be counted,

> and the trailing ` : i32` denotes the output type storing the count.

Reviewer: This is a lot of redundant basic MLIR for a `mlir-opt` tutorial! It's kind of confusing to me as to what the goal of this is at this point.

Author: I should have added this explanation to the PR description (I will copy this comment there for visibility).

The goal of this document is to provide the user the first thing they can do after building MLIR. As such, it is designed to get them started on their MLIR journey and hand-hold them through baby steps. It should be self-contained and have no strict dependencies on other documentation. IMO it's also acceptable that this tutorial differs slightly from the way an experienced person would do the same task. [A tutorial for beginners is not the same thing as a manual for best practice.](https://docs.divio.com/documentation-system/tutorials/#how-to-write-good-tutorials)

I want to demonstrate how to use `mlir-opt` to run both a lowering and a within-dialect optimization pass, and to explain the terminology required to demonstrate that. For example, showing a pass on the affine dialect requires me to explain, at least in brief, what the affine dialect is and why it exists. Without that sort of context, this would be less of a tutorial and more of a reference document (which can be good, but is not what I'm trying to do here). Toy Chapter 2 does better, but that still attempts to be exhaustive and spends a lot of time connecting the exhaustive description to MLIR's design rationale, whereas here I'm trying to explain the minimum possible to be able to read the programs in the examples. Also, I want this document to be meaningfully different from the help text of `mlir-opt`, in that it contains context not suitable for that help text, and also focuses on a very small subset of the available capability of `mlir-opt`. 

Generally speaking, I think it's much better to spend a paragraph or two to introduce a topic rather than ask the reader to go read a much longer reference spec before they can proceed. E.g., the MLIR language reference, which is written in a very comprehensive and generic fashion, introduces a few dozen concepts in its first four paragraphs, before giving a 60-line example program for a relatively complicated task that also depends in some sense on knowledge of TensorFlow.

It's also good to have tutorials with overlapping subsets of information.

    affine.store %cst, %1[%arg2] : memref<10xf32>
  }
  affine.for %arg2 = 0 to 10 {
    %2 = affine.load %0[%arg2] : memref<10xf32>
    %3 = arith.addf %2, %2 : f32
    affine.store %3, %arg0[%arg2] : memref<10xf32>
  }
  affine.for %arg2 = 0 to 10 {
    %2 = affine.load %1[%arg2] : memref<10xf32>
    %3 = arith.mulf %2, %2 : f32
    affine.store %3, %arg1[%arg2] : memref<10xf32>
  }
  return
}
```

This program introduces some additional dialects.
The [`affine` dialect](https://mlir.llvm.org/docs/Dialects/Affine/) mentioned in the introduction
represents well-structured loop nests,
and the [`affine.for` operation](https://mlir.llvm.org/docs/Dialects/Affine/#affinefor-affineaffineforop)
whose region corresponds to the loop's body.
`affine.for` also showcases some custom-defined syntax
to represent the loop bounds and loop induction variable.
The [`memref` dialect](https://mlir.llvm.org/docs/Dialects/MemRef/)
defines types and operations related to memory management
with pointer semantics.
Note also that while `memref` has store and load operations,
`affine` has its own that limit what types of memory accesses are allowed,
so as to ensure the well-structuredness of the loop nest.

> ## Running `mlir-opt`

Reviewer: Seems like here is the beginning of what I expect for a mlir-opt specific page.

After building the MLIR project,
the `mlir-opt` binary (located in `build/bin`)

> is the entry point for running passes and lowerings,

Reviewer: ```suggestion
is the entry point for running passes,
```

Lowerings aren't anything different than passes.

Author: I think it's worth introducing the concept distinctly, as the term is often used in the rest of the MLIR documentation.

as well as emitting debug and diagnostic data.

> Running `mlir-opt` with no flags will consume MLIR input

Reviewer: ```suggestion
Running `mlir-opt` with no flags will consume textual or bytecode IR
```

from standard in, parse and run verifiers on it,

> and write the MLIR back to standard out.

Reviewer: ```suggestion
and write the textual format back to standard out.
```

This is a good way to test if an input MLIR is well-formed.

`mlir-opt --help` shows a complete list of flags
(there are nearly 1000).
Each pass gets its own flag.

## Lowering `ctlz`

> Next we will show two of MLIR's lowering passes.

Reviewer: ```suggestion
Next we will show the execution of two MLIR passes with mlir-opt.
```

The first, `convert-math-to-llvm`, converts the `ctlz` op
to the `llvm` dialect's [`intr.ctlz` op](https://mlir.llvm.org/docs/Dialects/LLVM/#llvmintrctlz-llvmcountleadingzerosop)
which is an LLVM intrinsic.

> Note that `llvm` here is MLIR's `llvm` dialect,
> which would still need to be processed through `mlir-translate`
> to generate LLVM-IR.

Reviewer: ```suggestion
Note that `llvm` here is MLIR's `llvm` dialect.
```

`mlir-translate` looks like another chapter.

Recall our ctlz program:

```mlir
// mlir/test/Examples/mlir-opt/ctlz.mlir
func.func @main(%arg0: i32) -> i32 {
  %0 = math.ctlz %arg0 : i32
  func.return %0 : i32
}
```

After building MLIR, and from the `llvm-project` base directory, run

```bash

> build/bin/mlir-opt --convert-math-to-llvm mlir/test/Examples/mlir-opt/ctlz.mlir

Reviewer: ```suggestion
build/bin/mlir-opt --pass-pipeline=`builtin.module(convert-math-to-llvm)` mlir/test/Examples/mlir-opt/ctlz.mlir
```

The short syntax is error prone. Also that shows the need of documenting the pass nesting.

Author: I don't understand. How is it error prone?

Reviewer: Depending on the pass, it may just not run at all. 
Take for example this IR:
```
gpu.module @main_kernel_1 {
    gpu.func @main_kernel(%arg0: index, %arg1: index, %arg2: index, %arg3: memref<?x?x?xf32>, %arg4: memref<?x?x?xf32>) kernel attributes {gpu.known_block_size = array<i32: 1, 1, 1>} {
      gpu.return
    }
}
```

Now run: `bin/mlir-opt -test-gpu-memory-promotion /tmp/gpu.mlir` ; nothing will be executed. The correct invocation would be: `bin/mlir-opt --pass-pipeline="builtin.module(gpu.module(gpu.func(test-gpu-memory-promotion)))" /tmp/gpu.mlir`.

Author: What has bothered me about this is that there's no way to tell without looking at the source of the pass what the correct invocation is. I thought it would be safe for this tutorial to stick to example passes that anchor on builtin.module and have no surprises. What do you think?

Author: Given that the short syntax is used drastically more often in the MLIR codebase tests than the pass pipeline form, I think it is perhaps best to explain both, and then highlight this difference in the pass-pipeline section.

```
~/llvm-project/mlir/test
$ rg RUN | grep -- '--' | grep -v pass-pipeline | wc -l
1300

$ rg RUN | grep -- '--' | grep pass-pipeline | wc -l
133
```

Reviewer: > What has bothered me about this is that there's no way to tell without looking at the source of the pass what the correct invocation is,

The short syntax does not fix this aspect: it instead even making it worse! (the pass may not run and you will be left wondering why).

> Given that the short syntax is used drastically more often in the MLIR codebase tests than the pass pipeline form, 

This isn't a good reason: this is just that we haven't completed the migration after adding the pass-pipeline option and not being diligent in reviews to encourage the new syntax.

Author: > this is just that we haven't completed the migration

Does that imply the invocations without `--pass-pipeline` will be hard errors after the migration is complete? If they will remain legal, it seems reasonable to document them, especially considering that they account for most of the help text of the tool.

I am trying to quantify this a little bit. I only see a handful of passes that are neither using the default nor explicitly anchoring on `ModuleOp`:

```bash
~/fhe/llvm-project
> rg 'Pass : ' mlir/include/**/* | grep ", " | grep -v ModuleOp | grep -v "\"\"" | sort | uniq
mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:def GpuAsyncRegionPass : Pass<"gpu-async-region", "func::FuncOp"> {
mlir/include/mlir/Dialect/Linalg/Passes.td:def LinalgDetensorizePass : InterfacePass<"linalg-detensorize", "FunctionOpInterface"> {
mlir/include/mlir/Dialect/Tosa/Transforms/Passes.td:def TosaLayerwiseConstantFoldPass : Pass<"tosa-layerwise-constant-fold", "func::FuncOp"> {
mlir/include/mlir/Dialect/Vector/Transforms/Passes.td:def LowerVectorMaskPass : Pass<"lower-vector-mask", "func::FuncOp"> {
```

If it is just these four passes that can cause confusion, then it seems like we're imposing a lot of verbosity on the reader for not much benefit. IMO for a tutorial it is best to start with the simplest legal invocations first, and add in complexity later when it is needed. Are you concerned the confusion will not be clarified by the new `pass-pipeline` section, which explains the pitfall in more detail?

Reviewer: > Does that imply the invocations without --pass-pipeline will be hard errors after the migration is complete?

I don't think we'll ever deprecate and remove these.

>  If they will remain legal, it seems reasonable to document them, 

I would question whethr we have to document exhaustively all the variant and esoteric aspect of every API in a **tutorial** though?

> especially considering that they account for most of the help text of the tool.

That would be something else to fix :)

> I am trying to quantify this a little bit. I only see a handful of passes

I think you're grep is overly restrictive, try this one for instance:
```
$ git grep ': .*Pass<' mlir/include/**/* | grep "," | grep -v "\"ModuleOp" | grep -v "mlir::ModuleOp" |  sort   | wc -l
      63
```

> If it is just these four passes that can cause confusion, then it seems like we're imposing a lot of verbosity on the reader for not much benefit.

This isn't just about the confusion caused by the pass not running at all, it's also understanding the structure of the pass manager, the associated threading behavior, etc.

>  IMO for a tutorial it is best to start with the simplest legal invocations first

The "simplest" invocation is "barely legal" and **confusing**. The result of this is that users are forming a very incomplete mental model of the system, which lead them to be  puzzled by the behavior when they try to apply this to their project, and have hard time debugging what's happening.

This isn't unnecessary complexity.

Author: I disagree with most of what you wrote [1], but it doesn't seem I will convince you. I will change all `--pass` to `--pass-pipeline="builtin.module(pass)"` and tweak the prose to accommodate it.

Perhaps you should consider updating `mlir-opt` to emit a warning or hard error when one tries to use a non-`builtin.module`-anchored pass outside of `--pass-pipeline`.

[1]: except correcting my bad `git grep`, now I see about 50 passes, since yours also includes passes like `Pass<"nvvm-attach-target", "">` which I think anchor on module by default?

Reviewer: > Perhaps you should consider updating mlir-opt to emit a warning or hard error when one tries to use a non-builtin.module-anchored pass outside of --pass-pipeline.

Unfortunately I think we have a default nesting behavior that makes it implicitly work on immediately nested IR, that is if you have a `FuncOp` pass and use the short syntax it'll translate to `--pass-pipeline="builtin.module(func.func(passName))`
(this behavior is coming from our initial implementation of MLIR which had hard-coded module/func structure without any more nesting: there was never any ambiguity at the time).

```

which produces

```mlir
module {
  func.func @main(%arg0: i32) -> i32 {
    %0 = "llvm.intr.ctlz"(%arg0) <{is_zero_poison = false}> : (i32) -> i32
    return %0 : i32
  }
}
```

As you can see, the `math` dialect was the only thing that changed
due to the lowering.
The `func` dialect was left alone,
even though it also needs to be converted to the `llvm` dialect

> to generate LLVM-IR.

Reviewer: Can we just run `convert-to-llvm`?

Author: That is true, but it is a more advanced concept to explain. I can add a footnote describing how this is the actual way one would convert to LLVM.

Author: I'm convinced. Will update it.

What if the target machine does not have a `ctlz` intrinsic?
In this case, one can run the `--convert-math-to-funcs`
pass, which replaces the op with an implementation using

> other MLIR dialects.

Reviewer: I don't quite see why this paragraph is useful for understanding mlir-opt behavior.

```

```mlir
module {
  func.func @producer_consumer_fusion(%arg0: memref<10xf32>, %arg1: memref<10xf32>) {
    %alloc = memref.alloc() : memref<10xf32>
    %alloc_0 = memref.alloc() : memref<10xf32>
    %cst = arith.constant 0.000000e+00 : f32
    affine.for %arg2 = 0 to 10 {
      affine.store %cst, %alloc[%arg2] : memref<10xf32>
      affine.store %cst, %alloc_0[%arg2] : memref<10xf32>
    }
    affine.for %arg2 = 0 to 10 {
      %0 = affine.load %alloc[%arg2] : memref<10xf32>
      %1 = arith.addf %0, %0 : f32
      affine.store %1, %arg0[%arg2] : memref<10xf32>
    }
    affine.for %arg2 = 0 to 10 {
      %0 = affine.load %alloc_0[%arg2] : memref<10xf32>
      %1 = arith.mulf %0, %0 : f32
      affine.store %1, %arg1[%arg2] : memref<10xf32>
    }
    return
  }
}
```

Options passed to a pass
should come in the form of a quoted string
(to join all options into a single shell argument)
with space-separated `key=value` pairs.

## Building a pass pipeline on the command line

One can combine passes on the command line in two ways.

First, by simply placing the pass flags one after the other,
they will be run in order.

```bash

> build/bin/mlir-opt --convert-math-to-llvm --convert-func-to-llvm mlir/test/Examples/mlir-opt/ctlz.mlir

Reviewer: We should use `convert-to-llvm` in general, so we should find a better example.
Also we should only document `--pass-pipeline` syntax.
